components/persistor/lib/knex/db.ts

```diff
@@ -791,55 +791,89 @@ module.exports = function (PersistObjectTemplate) {
-                    else
-                        table[type](columns, name);
-
-                }));
+        var applyTableChanges = async function(dbChanges) {
+            function logTableChangesWarningMessage(columns, type, indexName, operation, error) {
+                const logger = PersistObjectTemplate && PersistObjectTemplate.logger;
+                if (logger) {
+                    logger.warn({
+                        function: 'syncIndexesForHierarchy',
+                        data: {
+                            type,
+                            columns,
+                            operation,
+                            indexName,
+                            tableName
+                        },
+                        module: 'db - applyTableChanges',
+                        error: error
+                    }, 'Executing one index at a time - Unable to update index');
+                }
             }
+            async function syncIndexesForHierarchy (operation, diffs) {
+                for (const _key in diffs[operation]) {
+                    try {
+                        const object = diffs[operation][_key];
+                        var type = object.def.type;
+                        var columns = object.def.columns;
+                        if (type !== 'unique' && type !== 'index') {
+                            throw new Error('index type can be only "unique" or "index"');
+                        }
 
+                        var name = _.reduce(object.def.columns, function (name, col) {
+                            return name + '_' + col;
+                        }, 'idx_' + tableName);
 
-            return knex.transaction(function (trx) {
-                return trx.schema.table(tableName, function (table) {
-                    syncIndexesForHierarchy('delete', dbChanges, table);
-                    syncIndexesForHierarchy('add', dbChanges, table);
-                    syncIndexesForHierarchy('change', dbChanges, table);
-                }).catch(function (error) {
-                    if (error && error.message && error.message === 'index type can be only "unique" or "index"') {
-                        throw error;
-                    }
-                    const logger = PersistObjectTemplate && PersistObjectTemplate.logger;
-                    if (logger) {
-                        logger.warn({
-                            component: 'persistor',
-                            dbChanges: dbChanges,
-                            module: 'db',
-                            activity: 'synchronizeIndexes',
-                            error: error
-                        }, 'Unable to update index');
+                        name = name.toLowerCase();
+                        if (operation === 'add') {
+                            try {
+                                await knex.schema.table(tableName, async function (table) {
+                                    await table[type](columns, name);
```

Comment:
are we sure the `table`being used here (passed in the callback) from `knex.schema.table` is async? we should verify that